File: 2015/python/d14.py
def read(filename):
    stats = {}
    for line in open(filename, 'r'):
        str = line.replace('\n', '').\
                   replace(' can fly', '').\
                   replace('km/s for ', '').\
                   replace('seconds, but then must rest for ', '').\
                   replace(' seconds.', '').split(' ')
        stats[str[0]] = (int(str[1]), int(str[2]), int(str[3]))
    return stats

def distances(stats, time):
    distances = [-1] * len(stats)
    i = 0
    for reindeer in stats:
        (speed, run, rest) = stats[reindeer]
        dist = int(time / (run + rest)) * speed * run
        leftover = time % (run + rest)
        distances[i] = dist + min(leftover, run) * speed
        i += 1
    return distances

def part1(stats, time = 2503):
    return max(distances(stats, time))

def part2(stats, time = 2503):
    scores = [0] * len(stats)
    for i in range(1, time + 1):
        dists = distances(stats, i)
        best = max(dists)
        for j in range(len(stats)):
            if dists[j] == best:
                scores[j] += 1
    return max(scores)

def tests():
    stats = read('../Java/d14/test.txt')
    assert(part1(stats, 1000) == 1120)
    assert(part2(stats, 1000) == 689)

tests()
stats = read('../inputs/input_14.txt')
print(part1(stats))
print(part2(stats))

File: 2015/R/d14.R
d <- read.csv("../inputs/input_14.txt", header = FALSE, sep = " ")[, c(1, 4, 7,14)]
names(d) <- c("reindeer", "speed", "time", "rest")

solve1 <- function(d, t) {
  max(((t %/% (d$time + d$rest)) * d$time * d$speed) + 
       pmin(d$time, (t %% (d$time + d$rest))) * d$speed)
}

solve2 <- function(d, t) {
  d$score <- 0
  for (i in seq_len(t)) {
    d$dist <- ((i %/% (d$time + d$rest)) * d$time * d$speed) + 
          pmin(d$time, (i %% (d$time + d$rest))) * d$speed
    d$score[d$dist == max(d$dist)] <- d$score[d$dist == max(d$dist)] + 1
  }
  max(d$score)
}

c(solve1(d, 2503), solve2(d, 2503))
